### DISCUSSION TITLE
something similar to useMutation from react-query

### DISCUSSION THREAD

**Discussion by bogdansoare** [vercel/swr#550] [General] [answered]
Is there something similar to the [useMutation](https://react-query.tanstack.com/docs/guides/mutations) hook from react-query or someway it can be replicated ?

Basically what I want is to have loading state while a POST request is happening

**Comment by sergiodxa** [comment, marked as answer]
You can use [use-mutation](https://github.com/sergiodxa/use-mutation), it will give you a similar API and is not dependent of RQ or SWR

**Reply by bogdansoare** [reply]
use-mutation looks great, I will start experimenting with it, thank you

**Comment by huozhi** [comment]
@bogdansoare you can also use react basic hooks to wrap on top of `mutate` API to build your own `useMutation`

### Implementation Example

```js
import { useCallback } from "react";
import { mutate } from "swr";

function useMutation(muationFn = f => f, options = {}) {
  const mutationCallback = useCallback(
    async (key, _payload) => {
      const {
        onSuccess = () => {},
        onMutate = () => {},
        onError = () => {}
      } = options;
      onMutate(key, _payload);
      try {
        const payload = await muationFn(_payload);
        await mutate(key, payload, false);
        onSuccess(key, payload);
      } catch (e) {
        onError(e);
      }
    },
    [muationFn, options]
  );
  return mutationCallback;
}

export default useMutation;
```

### Usage

```js
import React from "react";
import useSWR from "swr";
import useMutation from "./useMutation";

function fetcher() {
  return Promise.resolve(123);
}

export default function App() {
  const { data } = useSWR("key", fetcher, {});
  const mutateZero = useMutation();

  return (
    <div
      className="App"
      onClick={() => {
        mutateZero("key", 0);
      }}
    >
      <h1>Data: {data}</h1>
    </div>
  );
}
```

**Reply by bogdansoare** [reply]
cool usage of a custom hook, I will try it out, thank you

**Comment by ramyareye** [comment]
Recently I used [graphql-codegen-plugin-typescript-swr](https://github.com/croutonn/graphql-codegen-plugin-typescript-swr) and saw it uses [graphql-request](https://github.com/prisma-labs/graphql-request) under the hood!

**Comment by bryanprimus** [comment]
can someone provide an example of a custom useMutation hook in swr along with the type support with typescript?

**Reply by sergiodxa** [reply]
Check https://github.com/sergiodxa/use-mutation
